File: src/tools/pmd-cpd/eval-repos/synthetic/python/light_duplication.py
"""Python file with light duplication - small duplicated blocks."""


def process_user_data(user: dict) -> dict:
    """Process user data with validation."""
    result = {}
    result["id"] = user.get("id", 0)
    result["name"] = user.get("name", "").strip()
    result["email"] = user.get("email", "").lower().strip()
    result["active"] = user.get("active", True)
    result["created_at"] = user.get("created_at", "")
    return result


def process_admin_data(admin: dict) -> dict:
    """Process admin data with validation - light duplication."""
    result = {}
    result["id"] = admin.get("id", 0)
    result["name"] = admin.get("name", "").strip()
    result["email"] = admin.get("email", "").lower().strip()
    result["active"] = admin.get("active", True)
    result["created_at"] = admin.get("created_at", "")
    result["permissions"] = admin.get("permissions", [])
    return result


def validate_email(email: str) -> bool:
    """Validate email format."""
    if not email:
        return False
    if "@" not in email:
        return False
    parts = email.split("@")
    if len(parts) != 2:
        return False
    if not parts[0] or not parts[1]:
        return False
    if "." not in parts[1]:
        return False
    return True


def format_currency(amount: float, currency: str = "USD") -> str:
    """Format a number as currency."""
    symbols = {"USD": "$", "EUR": "E", "GBP": "P"}
    symbol = symbols.get(currency, currency)
    if amount < 0:
        return f"-{symbol}{abs(amount):,.2f}"
    return f"{symbol}{amount:,.2f}"


def calculate_discount(price: float, discount_pct: float) -> float:
    """Calculate discounted price."""
    if discount_pct < 0 or discount_pct > 100:
        raise ValueError("Discount must be between 0 and 100")
    return price * (1 - discount_pct / 100)

File: src/tools/pmd-cpd/eval-repos/synthetic/rust/light_duplication.rs
//! Rust file with light duplication - small duplicated blocks.

/// User data representation.
#[derive(Clone, Default)]
pub struct UserData {
    pub id: u32,
    pub name: String,
    pub email: String,
    pub active: bool,
    pub created_at: String,
    pub permissions: Vec<String>,
}

/// Process user data with validation.
pub fn process_user_data(user: &UserData) -> UserData {
    UserData {
        id: user.id,
        name: user.name.trim().to_string(),
        email: user.email.to_lowercase().trim().to_string(),
        active: user.active,
        created_at: user.created_at.clone(),
        permissions: Vec::new(),
    }
}

/// Process admin data with validation - light duplication.
pub fn process_admin_data(admin: &UserData) -> UserData {
    UserData {
        id: admin.id,
        name: admin.name.trim().to_string(),
        email: admin.email.to_lowercase().trim().to_string(),
        active: admin.active,
        created_at: admin.created_at.clone(),
        permissions: admin.permissions.clone(),
    }
}

/// Validate email format.
pub fn validate_email(email: &str) -> bool {
    if email.is_empty() {
        return false;
    }
    if !email.contains('@') {
        return false;
    }
    let parts: Vec<&str> = email.split('@').collect();
    if parts.len() != 2 {
        return false;
    }
    if parts[0].is_empty() || parts[1].is_empty() {
        return false;
    }
    if !parts[1].contains('.') {
        return false;
    }
    true
}

/// Format a number as currency.
pub fn format_currency(amount: f64, currency: &str) -> String {
    let symbol = match currency {
        "USD" => "$",
        "EUR" => "E",
        "GBP" => "P",
        _ => currency,
    };
    if amount < 0.0 {
        format!("-{}{:.2}", symbol, amount.abs())
    } else {
        format!("{}{:.2}", symbol, amount)
    }
}

/// Calculate discounted price.
pub fn calculate_discount(price: f64, discount_pct: f64) -> Result<f64, &'static str> {
    if discount_pct < 0.0 || discount_pct > 100.0 {
        return Err("Discount must be between 0 and 100");
    }
    Ok(price * (1.0 - discount_pct / 100.0))
}
